libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt

```diff
@@ -1534,6 +1476,146 @@ internal class MapRouteLine(
+         * road class section and associate that data with a traffic congestion value.
+         *
+         * Each item returned represents a distance from the origin point, the road class
+         * beginning at that distance and the traffic congestion value for that section. The
+         * road class and congestion remain until the next item in the collections.
+         *
+         * The items returned are ordered from the point closest to the origin to the point
+         * farthest from the origin. In other words from the beginning of the route until the end.
+         *
+         * @param route the the calculations should be performed on.
+         */
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.mapNotNull { it.steps() }
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    var runningDistance = 0.0
+                    val routeLineTrafficData = mutableListOf<RouteLineTrafficExpressionData>()
+                    route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.forEachIndexed { index, congestion ->
+                        runningDistance += distanceList[index]
+                        if (index == 0) {
+                            routeLineTrafficData.add(
+                                RouteLineTrafficExpressionData(
+                                    runningDistance,
+                                    congestion,
+                                    roadClassMap[index]
+                                )
+                            )
+                        } else {
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                routeLineTrafficData.last().roadClass == roadClassMap[index]
+                            ) {
+                                // continue
+                            } else if (routeLineTrafficData.last().roadClass != null &&
+                                roadClassMap[index] == null
+                            ) {
```

Comment:
This check is too lenient. In a scenario when the congestion level changes between intersections, we will skip and ignore it.

For example, running `getRouteLineExpressionDataWithStreetClassOverrideWhenHasStreetClassesOnMotorway` test and changing the congestion array from all `unknown`:
```
"congestion":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"]
```
to:
```
"congestion":["unknown","severe","severe","unknown","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"]
```
still lets the test pass. It ignores all of the severe congestion levels because they do not start on intersections geometry points.

Could you add another test case for this scenario?

Reply:
Added additional testing.